design/vela-core/definition-versioning.md

```diff
@@ -0,0 +1,296 @@
+
+**BDD Acceptance Criteria**
+
+>Scenario: Use the version specified in the Application manifest when deploying the service\
+>**GIVEN** a Component A with version 1.2.3\
+>**AND** a Component B with version 4.5.6\
+>**AND** an Application composed of A 1.2.2 and B 4.4.2\
+>**WHEN** the Application is deployed\
+>**THEN** it uses Component A 1.2.2 and B 4.4.2
+
+>**Variant:** Use the latest version for the part of the semVer that is not specified.\
+>**AND** an Application composed of A 1.2 and B 4\
+>**WHEN** the Application is deployed\
+>**THEN** it uses Component A 1.2.3 and B 4.5.6
+
+## Current Implementation
+
+Currently, Kubevela has some support for controlling Definition versions based on K8s annotations and DefinitionRevisions. The annotation `definitionrevision.oam.dev/name` can be used to version the ComponentDefinition. For example if the following annotation is added to a ComponentDefinition, it produces a new DefinitionRevision and names the ComponentDefinition as `component-name-v4.4` .
+
+> definitionrevision.oam.dev/name: "4.4"
+
+![version](./kubevela-version.png)
+
+This Component can then be referred in the Application as either of the following:
+
+>"component-name@v4.4" - `NamedDefinitionRevision`\
+>"component-name@v3"   - `DefinitionRevision`
+
+This versioning scheme, although convenient, has the following issues:
+
+- Since, there are more than one ways to reference a Component (`NamedDefinitionRevision` or `DefinitionRevision`), it proves difficult to maintain version consistency across clusters.
+- Each change in the ComponentDefintion Spec creates a new DefinitionRevision only if the annotation `definitionrevision.oam.dev/name` is not present at all. If there is change in the ComponentDefintion Spec when the annotation  `definitionrevision.oam.dev/name` is present and not updated, the DefinitionRevision is updated in place. This behavior adds to the inconsistency.
+
+#### Reference:
+
+https://kubevela.io/docs/platform-engineers/x-def-version/
+
+
+## Proposal 1
+
+### Details
+
+This implementation, suggested [here](https://github.com/kubevela/kubevela/issues/6435#issuecomment-1892372596), proposes adding another explicit annotation for keeping track of ComponentDefinition Revisions. For example, we can add an annotation like `definitionrevision.oam.dev/version` and use that to generate the ComponentDefinition Revisions.
+
+### Issues
+
+The following issues assume adherence to strict backward compatibility.
+- It might be very confusing for the users, because there would be no distinction between the Revisions generated by the `name` and `version` annotations. Additionally, since a new CRD API version is not created, Kubevela admins will need to keep track of the Components & Revisions created before and after the rollout of this feature.
+- If both the annotations  `definitionrevision.oam.dev/name` and `definitionrevision.oam.dev/version` are present, there is no clear way to decide the Revision behavior.
```

Comment:
We can add `version` field in definitions, and `version` will override `definitionrevision.oam.dev/name`